pass/pass.go

```diff
@@ -138,7 +138,7 @@ func (p Pass) Get(serverURL string) (string, string, error) {
 
 	if _, err := os.Stat(path.Join(getPassDir(), PASS_FOLDER, encoded)); err != nil {
 		if os.IsNotExist(err) {
-			return "", "", nil
+			return "", "", credentials.NewErrCredentialsNotFound()
 		}
 
 		return "", "", err
```

Comment:
Looks like this whole bock was explicitly added in https://github.com/docker/docker-credential-helpers/pull/84 (linked to https://github.com/docker/cli/pull/451).

I don't recall the specifics though; this may have been related to _other_ binaries named `pass` on Linux, which would get called, so the check for that was to see if a credentials store was initialised, and otherwise fall back to regular credentials storage 🤔

Reply:
The PR's description is:

> It turns out the cred helpers protocol is to return "" and not an error when a credential isn't present, so let's do that.

Is it possible the protocol changed since then?

I have a hard time reconciling that with the test. I didn't write this new test; I copied it from the other helpers:
https://github.com/docker/docker-credential-helpers/blob/master/wincred/wincred_windows_test.go#L238
https://github.com/docker/docker-credential-helpers/blob/master/osxkeychain/osxkeychain_darwin_test.go#L207
which confirm that this is supposed to return an error.

Reply:
maybe the API contract changed to return a standardized error instead of an implementation-specific one?

I also considered refactoring the tests into a "API invariants" test suite that we share across all the helpers. happy to do that in this PR (or in a follow-up one)

Reply:
I'm not aware of changes in the API, but must really admit that at the time I wasn't closely involved in development of the "credential-helpers".
 

Thinking out loud here; as that patch was opened related to https://github.com/docker/cli/pull/451, could it be that (as, unlike macOS and Windows) Linux had multiple possible helpers, that it was implemented as a "hack" to;

- check if the `pass` backend can be supported (don't error on "missing")
- and _only_ fall back to the second alternative (`secretservice` helper) otherwise
- and as last resort, fallback to "store inside the config-file"

Reply:
i think that's a reasonable hypothesis, but i don't think that's what's happening. The switching code is pretty straightforward. It calls Get() here:

https://github.com/docker/cli/blob/master/cli/config/credentials/native_store.go#L118

and the switch between pass and secretservice is here:

https://github.com/docker/cli/blob/master/cli/config/credentials/default_store_linux.go#L8

And as far as I can tell, `pass/pass.go` NEVER implemented this protocol correctly. Here's the first commit in August 2017:
https://github.com/docker/docker-credential-helpers/commit/1ab10377076c7cd77d4ee210dc101f04486e1fbe
even though the macos credential helper returned this error in at least February 2016:
https://github.com/docker/docker-credential-helpers/commit/4b716da07041687901f4ee4ab37134753965b7f9